File: ios/calorietracker/Medications/Models/MedicationDraft.swift
import Foundation

/// One validation error of the Add/Edit form (reference `validate_draft`, docs §15).
nonisolated struct DraftError: Hashable, Sendable, Identifiable {
    var field: String
    var code: String

    var id: String { field + ":" + code }

    /// Human copy for the inline footer of the form.
    var message: String {
        switch code {
        case "name_required": String(localized: "Enter a medicine name (up to 80 characters).")
        case "strength_too_long": String(localized: "Strength is too long (up to 40 characters).")
        case "form_invalid": String(localized: "Choose a form.")
        case "dose_quantity_invalid": String(localized: "Enter a dose greater than 0.")
        case "dose_unit_invalid": String(localized: "Choose a dose unit.")
        case "food_relation_invalid": String(localized: "Choose when to take it.")
        case "start_date_invalid": String(localized: "Choose a start date.")
        case "end_date_invalid": String(localized: "Choose a valid end date.")
        case "end_date_before_start": String(localized: "The end date is before the start date.")
        case "prn_has_schedule": String(localized: "As-needed medicines don't have a schedule.")
        case "frequency_required": String(localized: "Choose how often to take it.")
        case "times_required": String(localized: "Add at least one time.")
        case "times_invalid": String(localized: "Times must be unique (up to 12).")
        case "days_required": String(localized: "Choose at least one day.")
        case "days_invalid", "days_not_allowed": String(localized: "Check the selected days.")
        case "interval_invalid": String(localized: "Choose an interval that divides 24 hours.")
        case "anchor_required", "anchor_invalid": String(localized: "Choose the first dose time.")
        case "instructions_too_long": String(localized: "Instructions are too long (up to 200 characters).")
        case "note_too_long": String(localized: "The note is too long (up to 200 characters).")
        default: String(localized: "Check this field.")
        }
    }
}

/// Editable form model for Add / Edit / Import (docs §13, §15). Validation and the row shapes come
/// from the reference so the form can never save what the contract rejects.
nonisolated struct MedicationDraft: Hashable, Sendable {
    var name = ""
    var genericName = ""
    var brandName = ""
    var strength = ""
    var form: MedicationForm = .tablet
    var doseQuantity: Double? = 1
    var doseUnit: DoseUnit = .tablet
    var foodRelation: FoodRelation = .anytime
    var instructions = ""
    /// `yyyy-MM-dd`.
    var startDate: String
    /// `nil` = no end date.
    var endDate: String?
    var isPRN = false
    var frequency: ScheduleFrequency = .daily
    /// `HH:mm`; normalized (unique, ascending) before validation and saving.
    var times: [String] = ["08:00"]
    /// ISO weekdays 1 = Mon … 7 = Sun (weekly only).
    var days: [Int] = []
    var intervalHours = 8
    var anchorTime = "08:00"
    var reminderEnabled = true
    var relatedRecordID: String?
    /// A newly picked photo to store on save; `nil` keeps the current one.
    var newPhotoData: Data?
    /// Remove the stored photo on save.
    var removePhoto = false
    // Prescription-import metadata (informational).
    var hintConfidence: Double?
    var hintNotes: [String] = []
    var durationDays: Int?

    init(startDate: String) {
        self.startDate = startDate
    }

    /// Prefilled from an existing medication and its open schedule row.
    init(from medication: Medication, schedule: MedicationSchedule?) {
        startDate = medication.startDate
        name = medication.name
        genericName = medication.genericName ?? ""
        brandName = medication.brandName ?? ""
        strength = medication.strength ?? ""
        form = medication.form
        doseQuantity = medication.doseQuantity
        doseUnit = medication.doseUnit
        foodRelation = medication.foodRelation
        instructions = medication.instructions ?? ""
        endDate = medication.endDate
        isPRN = medication.isPRN
        relatedRecordID = medication.relatedRecordID
        if let schedule {
            frequency = schedule.frequency
            times = schedule.times.isEmpty ? ["08:00"] : schedule.times
            days = schedule.days
            intervalHours = schedule.intervalHours ?? 8
            anchorTime = schedule.anchorTime ?? "08:00"
            reminderEnabled = schedule.reminderEnabled
        }
    }

    /// Prefilled from a Records `medication` field through the reference's `frequency_hint` (docs §13).
    init(candidate: RecordMedicationValue, startDate: String, relatedRecordID: String?) {
        self.startDate = startDate
        self.relatedRecordID = relatedRecordID
        let value: RJ = .obj([
            "name": RJ.string(candidate.name), "strength": RJ.string(candidate.strength), "form": RJ.string(candidate.form),
            "dose": RJ.string(candidate.dose), "frequency": RJ.string(candidate.frequency),
            "duration": RJ.string(candidate.duration), "instructions": RJ.string(candidate.instructions),
        ])
        let hint = MR.frequencyHint(value)
        name = hint["name"].string ?? ""
        strength = hint["strength"].string ?? ""
        form = MedicationForm(raw: hint["form"].string)
        doseQuantity = hint["dose_quantity"].double
        doseUnit = DoseUnit(raw: hint["dose_unit"].string)
        foodRelation = FoodRelation(raw: hint["food_relation"].string)
        instructions = hint["instructions"].string ?? ""
        isPRN = hint["is_prn"].bool ?? false
        if !isPRN {
            frequency = ScheduleFrequency(raw: hint["frequency_kind"].string)
            let hintTimes = (hint["times"].array ?? []).compactMap(\.string)
            times = hintTimes.isEmpty ? ["08:00"] : hintTimes
            days = (hint["days"].array ?? []).compactMap { MR.int($0) }
            intervalHours = MR.int(hint["interval_hours"]) ?? 8
            anchorTime = hint["anchor_time"].string ?? "08:00"
        }
        hintConfidence = hint["confidence"].double
        hintNotes = (hint["notes"].array ?? []).compactMap(\.string)
        durationDays = MR.int(hint["duration_days"])
        if let durationDays, durationDays > 0, let start = MR.parseDate(startDate) {
            endDate = start.days(durationDays - 1).isoString
        }
    }

    // MARK: Normalization and validation

    /// Unique, ascending `HH:mm` values (the reference rejects unsorted or duplicate times).
    var normalizedTimes: [String] {
        Array(Set(times.filter { MR.parseHHMM($0) != nil })).sorted()
    }

    var normalizedDays: [Int] {
        Array(Set(days.filter { (1...7).contains($0) })).sorted()
    }

    private static func optional(_ s: String) -> RJ {
        let trimmed = s.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? .null : .str(trimmed)
    }

    /// The draft in the reference's `validate_draft` shape.
    var rj: RJ {
        var obj: [String: RJ] = [
            "name": .str(name.trimmingCharacters(in: .whitespacesAndNewlines)),
            "strength": Self.optional(strength),
            "form": .str(form.rawValue),
            "dose_quantity": RJ.number(doseQuantity),
            "dose_unit": .str(doseUnit.rawValue),
            "food_relation": .str(foodRelation.rawValue),
            "start_date": .str(startDate),
            "end_date": RJ.string(endDate),
            "is_prn": .bool(isPRN),
            "instructions": Self.optional(instructions),
            "note": .null,
        ]
        if isPRN {
            obj["frequency_kind"] = .null
            obj["times"] = .arr([])
            obj["days"] = .arr([])
            obj["interval_hours"] = .null
            obj["anchor_time"] = .null
        } else {
            obj["frequency_kind"] = .str(frequency.rawValue)
            obj["times"] = .arr(frequency == .interval ? [] : normalizedTimes.map(RJ.str))
            obj["days"] = .arr(frequency == .weekly ? normalizedDays.map(RJ.int) : [])
            obj["interval_hours"] = frequency == .interval ? .int(intervalHours) : .null
            obj["anchor_time"] = frequency == .interval ? .str(anchorTime) : .null
            obj["reminder_enabled"] = .int(reminderEnabled ? 1 : 0)
        }
        return .obj(obj)
    }

    var validationErrors: [DraftError] {
        MR.validateDraft(rj).compactMap { e in
            guard let field = e["field"].string, let code = e["code"].string else { return nil }
            return DraftError(field: field, code: code)
        }
    }

    var isValid: Bool { validationErrors.isEmpty }

    // MARK: Rows

    /// A new or updated `medications` row. `existing` keeps id, created_ms, status and photo.
    func medication(existing: Medication?, nowMs: Int64) -> Medication {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        func opt(_ s: String) -> String? {
            let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
            return t.isEmpty ? nil : t
        }
        return Medication(
            id: existing?.id ?? UUID().uuidString.lowercased(),
            name: trimmed,
            genericName: opt(genericName),
            brandName: opt(brandName),
            strength: opt(strength),
            form: form,
            doseQuantity: doseQuantity ?? 1,
            doseUnit: doseUnit,
            foodRelation: foodRelation,
            instructions: opt(instructions),
            startDate: startDate,
            endDate: endDate,
            status: existing?.status ?? .active,
            isPRN: isPRN,
            photoPath: existing?.photoPath,
            relatedRecordID: relatedRecordID,
            createdMs: existing?.createdMs ?? nowMs,
            updatedMs: nowMs
        )
    }

    /// The open schedule row for a non-PRN draft (`nil` for PRN). A NEW medication's row is active
    /// from `now − GRACE_MS` (docs §4 "creation", same on Android): a dose that became due within the
    /// last two hours still shows today and can be taken, while earlier slots of the day never existed
    /// (no phantom missed rows). Edits and resumes go through the reference's `lifecycle` (`now`).
    func schedule(medicationID: String, nowMs: Int64, zone: String) -> MedicationSchedule? {
        guard !isPRN else { return nil }
        return MedicationSchedule(
            medicationID: medicationID,
            frequency: frequency,
            times: frequency == .interval ? [] : normalizedTimes,
            days: frequency == .weekly ? normalizedDays : [],
            intervalHours: frequency == .interval ? intervalHours : nil,
            anchorTime: frequency == .interval ? anchorTime : nil,
            reminderEnabled: reminderEnabled,
            activeFromMs: nowMs - Int64(MR.graceMs),
            activeUntilMs: nil,
            createdMs: nowMs,
            updatedMs: nowMs
        )
    }

    /// `new_schedule` for the reference's `lifecycle('edit_schedule')`.
    var newScheduleRJ: RJ {
        .obj([
            "frequency_kind": .str(frequency.rawValue),
            "times": .arr(frequency == .interval ? [] : normalizedTimes.map(RJ.str)),
            "days": .arr(frequency == .weekly ? normalizedDays.map(RJ.int) : []),
            "interval_hours": frequency == .interval ? .int(intervalHours) : .null,
            "anchor_time": frequency == .interval ? .str(anchorTime) : .null,
            "reminder_enabled": .int(reminderEnabled ? 1 : 0),
        ])
    }

    /// True when the schedule part differs from `schedule` (frequency, times, days, interval, anchor).
    func scheduleDiffers(from schedule: MedicationSchedule?) -> Bool {
        guard let schedule else { return !isPRN }
        if frequency != schedule.frequency { return true }
        switch frequency {
        case .daily:
            return normalizedTimes != schedule.times
        case .weekly:
            return normalizedTimes != schedule.times || normalizedDays != schedule.days
        case .interval:
            return intervalHours != schedule.intervalHours || anchorTime != schedule.anchorTime
        }
    }
}

File: android/app/src/main/java/com/ayuvo/health/medications/model/MedicationDraft.kt
package com.ayuvo.health.medications.model

import com.ayuvo.health.medications.logic.MedicationConstants
import com.ayuvo.health.medications.logic.MedicationJson
import kotlinx.serialization.json.JsonObject

/**
 * The schedule half of the Add / Edit form (docs/medications.md §15). Validated as JSON by
 * `DraftValidation` so the rules stay identical to the reference.
 */
data class ScheduleDraft(
    val frequency: ScheduleFrequency? = ScheduleFrequency.DAILY,
    val times: List<String> = emptyList(),
    val days: List<Int> = emptyList(),
    val intervalHours: Int? = null,
    val anchorTime: String? = null,
    val reminderEnabled: Boolean = true
) {
    /** The `new_schedule` object of `lifecycle('edit_schedule')`. */
    fun toJson(): JsonObject = MedicationJson.obj(
        "frequency_kind" to frequency?.raw,
        "times" to times,
        "days" to days,
        "interval_hours" to intervalHours,
        "anchor_time" to anchorTime,
        "reminder_enabled" to if (reminderEnabled) 1 else 0
    )

    /**
     * The open row of a NEW medication (docs §4 "creation"): active from `now − GRACE_MS`, so a
     * dose that became due within the last two hours still shows today and can be taken, while
     * earlier slots of the day never existed (no phantom missed rows). Edits and resumes go
     * through the reference's `lifecycle`, which uses `now`.
     */
    fun toSchedule(id: String, medicationId: String, nowMs: Long): MedicationSchedule = MedicationSchedule(
        id = id,
        medicationId = medicationId,
        frequency = frequency ?: ScheduleFrequency.DAILY,
        times = if (frequency == ScheduleFrequency.INTERVAL) emptyList() else times,
        days = if (frequency == ScheduleFrequency.WEEKLY) days else emptyList(),
        intervalHours = if (frequency == ScheduleFrequency.INTERVAL) intervalHours else null,
        anchorTime = if (frequency == ScheduleFrequency.INTERVAL) anchorTime else null,
        reminderEnabled = reminderEnabled,
        activeFromMs = nowMs - MedicationConstants.GRACE_MS,
        activeUntilMs = null,
        createdMs = nowMs,
        updatedMs = nowMs
    )

    /** Same dosing rule as [other] (ids, versioning columns and the reminder flag ignored). */
    fun sameRuleAs(other: MedicationSchedule): Boolean =
        frequency == other.frequency && times == other.times && days == other.days &&
            intervalHours == other.intervalHours && anchorTime == other.anchorTime

    companion object {
        fun from(schedule: MedicationSchedule): ScheduleDraft = ScheduleDraft(
            frequency = schedule.frequency, times = schedule.times, days = schedule.days,
            intervalHours = schedule.intervalHours, anchorTime = schedule.anchorTime,
            reminderEnabled = schedule.reminderEnabled
        )
    }
}

/** The Add / Edit form state (docs §15). `validate_draft` runs on [toValidationJson]. */
data class MedicationDraft(
    val name: String = "",
    val genericName: String? = null,
    val brandName: String? = null,
    val strength: String? = null,
    val form: MedicationForm = MedicationForm.TABLET,
    val doseQuantity: Double? = 1.0,
    val doseUnit: DoseUnit = DoseUnit.TABLET,
    val foodRelation: FoodRelation = FoodRelation.ANYTIME,
    val instructions: String? = null,
    /** `yyyy-MM-dd`. */
    val startDate: String,
    val endDate: String? = null,
    val isPrn: Boolean = false,
    val photoPath: String? = null,
    val relatedRecordId: String? = null,
    val schedule: ScheduleDraft = ScheduleDraft()
) {
    /** The `draft` object of `validate_draft`. */
    fun toValidationJson(): JsonObject = MedicationJson.obj(
        "name" to name,
        "strength" to strength,
        "form" to form.raw,
        "dose_quantity" to doseQuantity,
        "dose_unit" to doseUnit.raw,
        "food_relation" to foodRelation.raw,
        "start_date" to startDate,
        "end_date" to endDate,
        "is_prn" to if (isPrn) 1 else 0,
        "frequency_kind" to if (isPrn) null else schedule.frequency?.raw,
        "times" to if (isPrn) emptyList() else schedule.times,
        "days" to if (isPrn) emptyList() else schedule.days,
        "interval_hours" to if (isPrn) null else schedule.intervalHours,
        "anchor_time" to if (isPrn) null else schedule.anchorTime,
        "instructions" to instructions
    )

    fun toMedication(id: String, nowMs: Long, status: MedicationStatus = MedicationStatus.ACTIVE, createdMs: Long = nowMs): Medication =
        Medication(
            id = id,
            name = name.trim(),
            genericName = genericName?.trim()?.takeIf { it.isNotEmpty() },
            brandName = brandName?.trim()?.takeIf { it.isNotEmpty() },
            strength = strength?.trim()?.takeIf { it.isNotEmpty() },
            form = form,
            doseQuantity = doseQuantity ?: 1.0,
            doseUnit = doseUnit,
            foodRelation = foodRelation,
            instructions = instructions?.trim()?.takeIf { it.isNotEmpty() },
            startDate = startDate,
            endDate = endDate,
            status = status,
            isPrn = isPrn,
            photoPath = photoPath,
            relatedRecordId = relatedRecordId,
            createdMs = createdMs,
            updatedMs = nowMs
        )

    fun toSchedule(id: String, medicationId: String, nowMs: Long): MedicationSchedule? =
        if (isPrn) null else schedule.toSchedule(id, medicationId, nowMs)

    companion object {
        fun from(medication: Medication, schedule: MedicationSchedule?): MedicationDraft = MedicationDraft(
            name = medication.name,
            genericName = medication.genericName,
            brandName = medication.brandName,
            strength = medication.strength,
            form = medication.form,
            doseQuantity = medication.doseQuantity,
            doseUnit = medication.doseUnit,
            foodRelation = medication.foodRelation,
            instructions = medication.instructions,
            startDate = medication.startDate,
            endDate = medication.endDate,
            isPrn = medication.isPrn,
            photoPath = medication.photoPath,
            relatedRecordId = medication.relatedRecordId,
            schedule = schedule?.let { ScheduleDraft.from(it) } ?: ScheduleDraft()
        )

        /** A draft from a Records prescription hint (§13); [startDate] is the device-local today. */
        fun fromHint(hint: FrequencyDraft, startDate: String, relatedRecordId: String?, endDateFor: (Int) -> String?): MedicationDraft =
            MedicationDraft(
                name = hint.name,
                strength = hint.strength,
                form = hint.form,
                doseQuantity = hint.doseQuantity,
                doseUnit = hint.doseUnit,
                foodRelation = hint.foodRelation,
                instructions = hint.instructions,
                startDate = startDate,
                endDate = hint.durationDays?.let(endDateFor),
                isPrn = hint.isPrn,
                relatedRecordId = relatedRecordId,
                schedule = if (hint.isPrn) ScheduleDraft() else ScheduleDraft(
                    frequency = hint.frequency ?: ScheduleFrequency.DAILY,
                    times = hint.times, days = hint.days,
                    intervalHours = hint.intervalHours, anchorTime = hint.anchorTime
                )
            )
    }
}
